simtools/simulator.py

```diff
@@ -0,0 +1,740 @@
+    printHistograms():
+        Print histograms and save a pdf file.
+    getListOfOutputFiles():
+        Get list of output files.
+    getListOfInputFiles():
+        Get list of input files.
+    getListOfLogFiles():
+        Get list of log files.
+    printListOfOutputFiles():
+        Print list of output files.
+    printListOfInputFiles():
+        Print list of output files.
+    printListOfLogFiles():
+        Print list of log files.
+    """
+
+    def __init__(
+        self,
+        label=None,
+        simulator=None,
+        filesLocation=None,
+        simtelSourcePath=None,
+        configData=None,
+        configFile=None,
+    ):
+        """
+        Simulator init.
+
+        Parameters
+        ----------
+        label: str
+            Instance label.
+        simulator: str
+            Type of simulator called (implemented is simtel and corsika).
+        filesLocation: str or Path.
+            Location of the output files. If not given, it will be set from \
+            the config.yml file.
+        simtelSourcePath: str or Path
+            Location of source of the sim_telarray/CORSIKA package.
+        configData: dict
+            Dict with shower or array model configuration data.
+        configFile: str or Path
+            Path to yaml file containing configurable data.
+        """
+        self._logger = logging.getLogger(__name__)
+        self._logger.debug("Init Simulator {}".format(simulator))
+
+        self.label = label
+        self._set_simulator(simulator)
+        self.runs = list()
+        self._results = defaultdict(list)
+
+        self._set_file_locations(filesLocation, simtelSourcePath)
+        self._loadConfigData(configData, configFile)
+
+        self._setSimulationRunner()
+
+    def _set_simulator(self, simulator):
+        """
+        Set and test simulator type
+
+        Parameters
+        ----------
+        simulator: str
+            ype of simulator called (implemented is simtel and corsika).
+
+        Raises
+        ------
+        gen.InvalidConfigData
+
+        """
+
+        self.simulator = simulator.lower()
+        if self.simulator not in ['simtel', 'corsika']:
+            raise gen.InvalidConfigData
+
+    def _set_file_locations(self, filesLocation=None, simtelSourcePath=None):
+        """
+        Set file locations, input and output directories
+
+        Parameters
+        ----------
+        filesLocation: str or Path.
+            Location of the output files. If not given, it will be set from \
+            the config.yml file.
+        simtelSourcePath: str or Path
+            Location of source of the sim_telarray/CORSIKA package.
+
+        """
+
+        self._simtelSourcePath = Path(cfg.getConfigArg("simtelPath", simtelSourcePath))
+        self._filesLocation = Path(cfg.getConfigArg("outputLocation", filesLocation))
+        try:
+            self._outputDirectory = io.getOutputDirectory(
```

Comment:
I am a bit surprised you decided to do it this way. 

So far all of these settings of output locations were done in `io_handler.py`. Including the CORSIKA and array-simulator which you left intact above. Instead, in this case, you decided to put this code here and call the function directly. I don't necessarily mind it, it's still fairly short, but then we are inconsistent with our use of `io_handler`. In particular, you might want to remove the special functions for CORSIKA and array-simulator.

Also, the exception for the FileNotFound is already done in `getOutputDirectoy`, not sure why you need it here as well.

Reply:
Agree on the exceptions - I added it first here and then later in the io_handler.

And I didn't get the other point, which special function do you mean here?

Reply:
What I mean is that in `io_handler` we have functions called `getCorsikaOutputDirectory` and `getArraySimulatorOutputDirectory`. These essentially do what you do here, they build the output directory by calling `getOutputDirectory` and providing the label and file location. I am surprised you didn't use them (and maybe added one for sim_telarray) and instead decided to make `getOutputDirectory` public and added this code here. Unless there's a good reason for this, I really recommend changing it and using `io_handler` like we do everywhere else.

Reply:
Using getCorsikaOutputDirectory and getArraySimulatorOutputDirectory would mean to add ifs to distinguish between the simulation codes. I don't want to do it, it is messy and will get worse with more simulation types

Reply:
I don't see why one `if` (or 2-3 in the future) is worse than changing the whole philosophy of `io_handler`. In addition, the `if` can be placed in `io_handler` rather than here. However, I will not insist on this point. If you prefer doing it this way, that's fine.